File: .claude/skills/wikimedia-eventstreams/scripts/sse-relay-server.py
#!/usr/bin/env python3
"""
SSE Relay: EventStreams → Your Server → Browser Clients (Python/FastAPI)

Consumes Wikimedia EventStreams (recentchange) and relays filtered events
to browser clients via a secondary SSE endpoint.

Usage:
    pip install fastapi uvicorn aiohttp aiohttp-sse-client
    python scripts/sse-relay-server.py

Then open http://localhost:8000/events in a browser.
"""

import asyncio
import json

import aiohttp
from aiohttp_sse_client import client as sse_client
from fastapi import FastAPI, Request
from fastapi.responses import StreamingResponse

app = FastAPI()
clients = []


async def consume_eventstream():
    headers = {"User-Agent": "MyApp/1.0 (me@example.com) LiveDashboard"}
    async with aiohttp.ClientSession(headers=headers) as session:
        async with sse_client.EventSource(
            "https://stream.wikimedia.org/v2/stream/recentchange",
            session=session,
        ) as event_source:
            async for event in event_source:
                if event.type == "message":
                    change = json.loads(event.data)
                    if change.get("meta", {}).get("domain") == "canary":
                        continue
                    # Customize this filter to your needs
                    # For example, Commons uploads:
                    # if change.get("wiki") == "commonswiki" and ...
                    for q in clients:
                        await q.put(json.dumps(change))


@app.get("/events")
async def sse_endpoint(request: Request):
    queue: asyncio.Queue = asyncio.Queue()
    clients.append(queue)

    async def generate():
        try:
            yield "event: connected\ndata: {}\n\n"
            while True:
                data = await queue.get()
                yield f"event: update\ndata: {data}\n\n"
        except asyncio.CancelledError:
            clients.remove(queue)

    return StreamingResponse(generate(), media_type="text/event-stream")


@app.on_event("startup")
async def startup():
    asyncio.create_task(consume_eventstream())


if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000)

File: .claude/skills/wikimedia-eventstreams/scripts/sse-relay-server.js
#!/usr/bin/env node
/**
 * SSE Relay: EventStreams → Your Server → Browser Clients
 *
 * Consumes Wikimedia EventStreams (recentchange) and relays filtered events
 * to browser clients via a secondary SSE endpoint.
 *
 * Usage:
 *   npm install express eventsource
 *   node scripts/sse-relay-server.js
 *
 * Then open http://localhost:3000 in a browser and connect to /events.
 */

const express = require('express');
const { EventSource } = require('eventsource');

const PORT = process.env.PORT || 3000;
const USER_AGENT = 'MyApp/1.0 (me@example.com) LiveDashboard';

// ── Connected browser clients ──
const clients = new Set();

function broadcast(data) {
    const msg = `event: update\ndata: ${JSON.stringify(data)}\n\n`;
    for (const client of clients) {
        try { client.write(msg); } catch { clients.delete(client); }
    }
}

// ── Consume EventStreams (single upstream connection) ──
const es = new EventSource(
    'https://stream.wikimedia.org/v2/stream/recentchange',
    { headers: { 'User-Agent': USER_AGENT } }
);

es.onmessage = (event) => {
    const data = JSON.parse(event.data);
    if (data.meta?.domain === 'canary') return;
    // Customize this filter to your needs
    if (data.wiki === 'enwiki' && data.type === 'edit' && !data.bot) {
        broadcast({ user: data.user, title: data.title, wiki: data.wiki });
    }
};

es.onerror = (err) => console.error('EventStreams error:', err.message);

// ── SSE endpoint for browser clients ──
const app = express();

app.get('/events', (req, res) => {
    res.writeHead(200, {
        'Content-Type': 'text/event-stream',
        'Cache-Control': 'no-cache',
        'Connection': 'keep-alive',
    });
    clients.add(res);
    res.write('event: connected\ndata: {}\n\n');

    req.on('close', () => {
        clients.delete(res);
        console.log(`[SSE] Client disconnected (${clients.size} remaining)`);
    });
});

app.listen(PORT, () => {
    console.log(`\n  🌐 SSE Relay running at http://localhost:${PORT}`);
    console.log(`  📡 SSE endpoint: http://localhost:${PORT}/events\n`);
});
